### PR TITLE
Updated

### PR DESCRIPTION
**Pull request by manas-143**

### CODE REVIEW

File: amazon_features/steps/amazon.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+        "ADD TO CART BTN": "xpath=//*[@id='add-to-cart-button']",
+        "CART ICON": "xpath=//*[@class='nav-cart-icon nav-sprite']",
+        "TOTAL CART PRICE": "xpath=//*[@id='sc-subtotal-amount-buybox']/span",
+        "ITEM PRICE": "xpath=//span[@id='tp_price_block_total_price_ww']/descendant::span[@class='a-price-whole']"
+        }
+""".............................................................."""
+
+
+@given(u'User is on the amazon website search for "{search_query}"')
+def step_impl(context, search_query):
```
Comment (Praveenkumar-95) [inline amazon_features/steps/amazon.py]: Please change the step name

Comment (manas-143) [reply amazon_features/steps/amazon.py]: ok praveen

File: amazon_features/steps/amazon.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
+from behave import given, when, then
+
+from features.steps.functions import send_input, click, select_items_extract_prices, total_cart_prices
+
+"""......................all the locators......................"""
+
+tags = {"SEARCH BOX": "xpath=//*[@id='twotabsearchtextbox']",
+        "SEARCH": "xpath=//*[@id='nav-search-submit-button']",
+        "RATING": "xpath =(//ul[@data-csa-c-content-id='1318475031']/span)",
```
Comment (Praveenkumar-95) [inline amazon_features/steps/amazon.py]: Are you sure these ID will remain same

File: amazon_features/steps/amazon.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+@given(u'User is on the amazon website search for "{search_query}"')
+def step_impl(context, search_query):
+    send_input(context.page, tags["SEARCH BOX"], search_query)  # user entered search item and searched for product
+    click(context.page, tags["SEARCH"])
+
+
+@when(u'User filter by "{star}" ratings')
+def filtering_the_items(context, star):
+    Dict = {"4": 1, "3": 2, "2": 3, "1": 4}  # created a dict to select the locator according to the user's input
```
Comment (Praveenkumar-95) [inline amazon_features/steps/amazon.py]: What is the purpose of the dictionary here ?,
can't we use the rating number as XPath ?

File: amazon_features/steps/functions.py [hunk 4]
```diff
@@ -0,0 +1,28 @@
+def send_input(page, selector, search_item):  # function for sending input to  search box
+    page.locator(selector).fill(search_item)
+
+
+def click(page, selector):  # click function to perform click
+    page.locator(selector).click()
+
+
+def select_items_extract_prices(page, selector1, selector2, selector3, number):  # function for item's price extract
```
Comment (Praveenkumar-95) [inline amazon_features/steps/functions.py]: Please provide the correct argument names to be passed for this function

File: amazon_features/steps/amazon.py (hunk 2 above)
Comment (manas-143) [reply amazon_features/steps/amazon.py]: I am not sure I just took as it is easy to find out the locator. suggest me what else I can do

Comment (Praveenkumar-95) [reply amazon_features/steps/amazon.py]: try to find out generic XPath to identify the rantings 
use can use **aria-label="3 Stars & Up"**

File: amazon_features/steps/amazon.py [hunk 5]
```diff
@@ -0,0 +1,53 @@
+from behave import given, when, then
+
+from functions import send_input, click, select_items_extract_prices, total_cart_prices
+
+"""......................all the locators......................"""
+
+tags = {
+    "SEARCH BOX": "xpath=//input[@type='text']",
+    "4 RATING": "xpath =//section[@aria-label='4 Stars & Up']",
```
Comment (Praveenkumar-95) [inline amazon_features/steps/amazon.py]: Can you use only one xpath for all the star ratings?

File: features/steps/google.py [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+
+
+@when(u'add top "{number}" places with its information')
+def adding_places_with_information(context, number):
+    context.L = []  # List of all places
+    num = int(number)  # Converting string to int
+    total_item = scrolling(context.page, tags["SEARCH ITEMS"], num)  # this returns the total number of items
+
+    for i in range(num):
+        D = {}  # adding the place details in dictionary
```
Comment (Praveenkumar-95) [inline features/steps/google.py]: variables should be small case

File: features/steps/amazon.py [hunk 7]
```diff
@@ -1,41 +1,53 @@
...
-from features.pages.HomePage import HomePage  # imported home page module to perform actions in homepage
-from features.pages.ItemListPage import ItemLists  # imported Item list page module to perform item sorting
-from features.pages.SelectItemPage import SelectItem  # imported select item page module to select the items
-from features.pages.CheckOutPage import CheckOut  # imported check out page module to compare the value
+from functions import send_input, click, select_items_extract_prices, total_cart_prices
 
+"""......................all the locators......................"""
 
-amount = []  # to store the price of each item
+tags = {
```
Comment (Praveenkumar-95) [inline features/steps/amazon.py]: change the variable name

File: features/steps/functions.py [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+    try:
+        text = page.locator(selector).text_content()
+    except:
+        text = "content not found"
+    return text
+
+
+def scrolling(page, selector, number):
+    while True:
+        all_item = page.query_selector_all(selector)  # all displayed items
```
Comment (Praveenkumar-95) [inline features/steps/functions.py]: Change the logic to catch the condition

File: features/steps/amazon.py [hunk 9]
```diff
@@ -1,41 +1,53 @@
...
+    "SEARCH BOX": "xpath=//input[@type='text']",
+    "4 RATING": "xpath =//section[@aria-label='4 Stars & Up']",
+    "3 RATING": "xpath=//section[@aria-label='3 Stars & Up']",
+    "2 RATING": "xpath=//section[@aria-label='2 Stars & Up']",
+    "1 RATING": "xpath=//section[@aria-label='1 Stars & Up']",
```
Comment (Praveenkumar-95) [inline features/steps/amazon.py]: Use string formattings

**Review by Praveenkumar-95 — CHANGES_REQUESTED** [review]

File: features/steps/functions.py [hunk 10]
```diff
@@ -0,0 +1,47 @@
+def send_input(page, selector, search_item):  # function for sending input to  search box
+    page.locator(selector).fill(search_item)
+
+
+def click(page, selector):  # click function to perform click
+    page.locator(selector).click()
+
+
+def select_items_extract_prices(page, item_list_selector, price_selector, cart_btn_selector,
```
Comment (Praveenkumar-95) [inline features/steps/functions.py]: thus function is specific to amazon...

File: features/steps/functions.py [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+            all_items.nth(i).click()  # click on each laptop
+        page1 = page1_info.value  # going to the next tab
+        amt = page1.locator(price_selector).inner_text()  # locator for amount
+        amount.append(float(amt.replace(",", "")))  # storing each laptop value in float
+        page1.locator(cart_btn_selector).click()  # adding the laptop to cart
+        page1.close()  # closing the new tab
+    return amount  # returning the amount to base page
+
+
+def total_cart_prices(page, selector):
```
Comment (Praveenkumar-95) [inline features/steps/functions.py]: thus function is specific to amazon...

File: features/steps/functions.py [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+    price = final_price.replace(",", "")
+    cart_amt = float(price)
+    return cart_amt
+
+
+def inner_text(page, selector):
+    try:
+        text = page.locator(selector).text_content()
+    except Exception as e:
+        text = e
```
Comment (Praveenkumar-95) [inline features/steps/functions.py]: Why the error is saved in a variable?

File: features/steps/google.py [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+
+        # adding individual details to list by removing duplicates
+        if item_dict not in context.L:
+            context.L.append(item_dict)
+
+
+@then(u'User makes a csv file to save the information')
+def saves_company_details_to_csv(context):
+    # Specify the CSV file path
+    csv_file_path = f'{context.search}'
```
Comment (Praveenkumar-95) [inline features/steps/google.py]: csv Extention is missing

File: features/steps/amazon.py [hunk 14]
```diff
@@ -1,41 +1,45 @@
...
-from features.pages.SelectItemPage import SelectItem  # imported select item page module to select the items
-from features.pages.CheckOutPage import CheckOut  # imported check out page module to compare the value
+from functions import send_input, click, select_items_extract_prices, total_cart_prices
 
+"""......................all the locators......................"""
 
-amount = []  # to store the price of each item
+locators = {
+    "SEARCH BOX": "xpath=//input[@type='text']",
+    "RATING": "xpath =//section[@aria-label='4 Stars & Up']",
```
Comment (Praveenkumar-95) [inline features/steps/amazon.py]: why the static rating 4 is mentioned?

File: features/steps/amazon.py [hunk 15]
```diff
@@ -1,41 +1,45 @@
...
+    context.page.keyboard.press("Enter")
 
 
-@when(u'User filter by ratings')
-def filtering_the_items(context):
-    item_list = ItemLists(context.page)
-    item_list.click()
+@when(u'User filter by "{star}" ratings')
+def filtering_the_items(context, star):
+    locators["RATING"] = f"xpath =//section[@aria-label='{star} Stars & Up']"
```
Comment (Praveenkumar-95) [inline features/steps/amazon.py]: Why it is re-assigned the variable of the Rating?

**Review by Praveenkumar-95 — CHANGES_REQUESTED** [review]

File: features/steps/Google_functions.py
Comment (Praveenkumar-95) [inline features/steps/Google_functions.py:1]: Is this specific to Google  ?

**Review by Praveenkumar-95 — CHANGES_REQUESTED** [review]

**Pull request merged**